File: node/src/bloom/bloom.js
/**
 * Simple bloom filter implementation for file-level column filtering.
 * Each parquet file gets one bloom filter per indexed column.
 * Uses Kirsch-Mitzenmacher double hashing: h(i) = h1 + i*h2
 */

const MIN_SIZE = 1024; // minimum bits to avoid degenerate small filters

function fnv1a(str) {
  let hash = 0x811c9dc5;
  for (let i = 0; i < str.length; i++) {
    hash ^= str.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash;
}

function murmurMix(str) {
  let h = 0x12345678;
  for (let i = 0; i < str.length; i++) {
    h ^= str.charCodeAt(i);
    h = Math.imul(h, 0x5bd1e995);
    h ^= h >>> 15;
  }
  return h >>> 0;
}

export class BloomFilter {
  constructor(size, hashCount) {
    this.size = size;
    this.hashCount = hashCount;
    this.bits = new Uint8Array(Math.ceil(size / 8));
  }

  add(value) {
    const str = String(value);
    const h1 = fnv1a(str);
    const h2 = murmurMix(str);
    for (let i = 0; i < this.hashCount; i++) {
      const bit = (h1 + i * h2) % this.size;
      this.bits[bit >> 3] |= 1 << (bit & 7);
    }
  }

  probe(value) {
    const str = String(value);
    const h1 = fnv1a(str);
    const h2 = murmurMix(str);
    for (let i = 0; i < this.hashCount; i++) {
      const bit = (h1 + i * h2) % this.size;
      if (!(this.bits[bit >> 3] & (1 << (bit & 7)))) {
        return false; // definitely absent
      }
    }
    return true; // maybe present
  }

  serialize() {
    return {
      size: this.size,
      hashCount: this.hashCount,
      bits: Buffer.from(this.bits).toString('base64'),
    };
  }

  static deserialize(data) {
    const bf = new BloomFilter(data.size, data.hashCount);
    bf.bits = new Uint8Array(Buffer.from(data.bits, 'base64'));
    return bf;
  }

  /**
   * Build a bloom filter from an array of values.
   * Auto-sizes based on expected element count for ~1% false positive rate.
   */
  static build(values, { falsePositiveRate = 0.01 } = {}) {
    const n = values.length || 1;
    const computed = Math.ceil((-n * Math.log(falsePositiveRate)) / (Math.LN2 * Math.LN2));
    const size = Math.max(MIN_SIZE, computed);
    const hashCount = Math.max(1, Math.round((size / n) * Math.LN2));
    const bf = new BloomFilter(size, hashCount);
    for (const v of values) {
      bf.add(v);
    }
    return bf;
  }
}

/**
 * Build file-level bloom filters for multiple columns.
 * Returns { columns: { colName: serialized bloom, ... } }
 */
export function buildFileBloom(columnValues) {
  const columns = {};
  for (const [col, values] of Object.entries(columnValues)) {
    const bf = BloomFilter.build(values);
    columns[col] = bf.serialize();
  }
  return { columns };
}

/**
 * Probe a file-level bloom for a set of equality filters.
 * Returns true if the file might contain matches, false if it definitely doesn't.
 */
function escapeSql(str) {
  return str.replace(/'/g, "''");
}

/**
 * Store a file-level bloom (from cold storage JSON) into DuckDB's _bloom_cache table.
 * Stores raw bytes as BLOB instead of base64.
 */
export async function storeBloomToDb(conn, tenant, segment, bloomData) {
  for (const [col, data] of Object.entries(bloomData.columns)) {
    const bits = Buffer.from(data.bits, 'base64');
    const hex = bits.toString('hex');
    await conn.run(
      `INSERT OR REPLACE INTO _bloom_cache VALUES ('${escapeSql(tenant)}', '${escapeSql(segment)}', '${escapeSql(col)}', ${data.size}, ${data.hashCount}, '\\x${hex}'::BLOB)`
    );
  }
}

/**
 * Load a file-level bloom from DuckDB's _bloom_cache table.
 * Returns the same { columns: { col: { size, hashCount, bits } } } format as JSON files,
 * or null if not found.
 */
export async function loadBloomFromDb(conn, tenant, segment) {
  const result = await conn.runAndReadAll(
    `SELECT column_name, filter_size, hash_count, encode(bits) AS bits_b64 FROM _bloom_cache WHERE tenant = '${escapeSql(tenant)}' AND segment = '${escapeSql(segment)}'`
  );
  const rows = result.getRowObjectsJson();
  if (rows.length === 0) return null;

  const columns = {};
  for (const row of rows) {
    columns[row.column_name] = {
      size: Number(row.filter_size),
      hashCount: Number(row.hash_count),
      bits: row.bits_b64,
    };
  }
  return { columns };
}

export function probeFileBloom(bloomData, filters) {
  for (const [col, value] of Object.entries(filters)) {
    if (!bloomData.columns[col]) continue; // no bloom for this column, can't skip
    const bf = BloomFilter.deserialize(bloomData.columns[col]);
    if (Array.isArray(value)) {
      // IN filter — file can be skipped only if ALL values are absent
      const anyPresent = value.some((v) => bf.probe(v));
      if (!anyPresent) return false;
    } else {
      if (!bf.probe(value)) return false;
    }
  }
  return true; // might contain matches
}

File: go/internal/bloom/bloom.go
package bloom

import (
	"database/sql"
	"encoding/base64"
	"fmt"
	"math"
)

const minSize = 1024

func fnv1a(s string) uint32 {
	h := uint32(0x811c9dc5)
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= 0x01000193
	}
	return h
}

func murmurMix(s string) uint32 {
	h := uint32(0x12345678)
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= 0x5bd1e995
		h ^= h >> 15
	}
	return h
}

type Filter struct {
	Size      int
	HashCount int
	Bits      []byte
}

func NewFilter(size, hashCount int) *Filter {
	if size < 1 {
		size = minSize
	}
	if hashCount < 1 {
		hashCount = 1
	}
	return &Filter{Size: size, HashCount: hashCount, Bits: make([]byte, int(math.Ceil(float64(size)/8.0)))}
}

func (f *Filter) Add(value string) {
	h1 := fnv1a(value)
	h2 := murmurMix(value)
	for i := 0; i < f.HashCount; i++ {
		bit := int((h1 + uint32(i)*h2) % uint32(f.Size))
		f.Bits[bit>>3] |= 1 << (bit & 7)
	}
}

func (f *Filter) Probe(value string) bool {
	h1 := fnv1a(value)
	h2 := murmurMix(value)
	for i := 0; i < f.HashCount; i++ {
		bit := int((h1 + uint32(i)*h2) % uint32(f.Size))
		if f.Bits[bit>>3]&(1<<(bit&7)) == 0 {
			return false
		}
	}
	return true
}

type Serialized struct {
	Size      int    `json:"size"`
	HashCount int    `json:"hashCount"`
	Bits      string `json:"bits"`
}

type FileBloom struct {
	Columns map[string]Serialized `json:"columns"`
}

func Deserialize(s Serialized) (*Filter, error) {
	bits, err := base64.StdEncoding.DecodeString(s.Bits)
	if err != nil {
		return nil, err
	}
	return &Filter{Size: s.Size, HashCount: s.HashCount, Bits: bits}, nil
}

func ParseFileBloom(raw map[string]any) (FileBloom, error) {
	colsAny, ok := raw["columns"].(map[string]any)
	if !ok {
		return FileBloom{}, fmt.Errorf("missing columns")
	}
	cols := make(map[string]Serialized, len(colsAny))
	for col, v := range colsAny {
		m, ok := v.(map[string]any)
		if !ok {
			continue
		}
		s, okS := m["size"].(float64)
		h, okH := m["hashCount"].(float64)
		b, okB := m["bits"].(string)
		if !okS || !okH || !okB {
			continue
		}
		cols[col] = Serialized{Size: int(s), HashCount: int(h), Bits: b}
	}
	return FileBloom{Columns: cols}, nil
}

func ProbeFileBloom(b FileBloom, filters map[string]any) bool {
	for col, value := range filters {
		ser, ok := b.Columns[col]
		if !ok {
			continue
		}
		bf, err := Deserialize(ser)
		if err != nil {
			continue
		}
		switch v := value.(type) {
		case []string:
			anyPresent := false
			for _, item := range v {
				if bf.Probe(item) {
					anyPresent = true
					break
				}
			}
			if !anyPresent {
				return false
			}
		case []any:
			anyPresent := false
			for _, item := range v {
				if bf.Probe(fmt.Sprintf("%v", item)) {
					anyPresent = true
					break
				}
			}
			if !anyPresent {
				return false
			}
		default:
			if !bf.Probe(fmt.Sprintf("%v", v)) {
				return false
			}
		}
	}
	return true
}

func escapeSQL(s string) string {
	out := make([]rune, 0, len(s))
	for _, r := range s {
		if r == '\'' {
			out = append(out, '\'', '\'')
			continue
		}
		out = append(out, r)
	}
	return string(out)
}

func StoreToDB(conn *sql.DB, tenant, segment string, b FileBloom) error {
	for col, data := range b.Columns {
		bits, err := base64.StdEncoding.DecodeString(data.Bits)
		if err != nil {
			return err
		}
		_, err = conn.Exec(
			`INSERT OR REPLACE INTO _bloom_cache VALUES (?, ?, ?, ?, ?, ?)`,
			tenant, segment, col, data.Size, data.HashCount, bits,
		)
		if err != nil {
			return err
		}
	}
	return nil
}

func LoadFromDB(conn *sql.DB, tenant, segment string) (FileBloom, bool, error) {
	rows, err := conn.Query(
		`SELECT column_name, filter_size, hash_count, bits FROM _bloom_cache WHERE tenant = ? AND segment = ?`,
		tenant, segment,
	)
	if err != nil {
		return FileBloom{}, false, err
	}
	defer rows.Close()

	cols := map[string]Serialized{}
	for rows.Next() {
		var col string
		var size int
		var hashCount int
		var bits []byte
		if err := rows.Scan(&col, &size, &hashCount, &bits); err != nil {
			return FileBloom{}, false, err
		}
		cols[col] = Serialized{Size: size, HashCount: hashCount, Bits: base64.StdEncoding.EncodeToString(bits)}
	}
	if err := rows.Err(); err != nil {
		return FileBloom{}, false, err
	}
	if len(cols) == 0 {
		return FileBloom{}, false, nil
	}
	return FileBloom{Columns: cols}, true, nil
}
